File: internal/model/site_channel.go
package model

type SiteChannelCard struct {
	SiteID       int                  `json:"site_id"`
	SiteName     string               `json:"site_name"`
	BaseURL      string               `json:"base_url"`
	Platform     SitePlatform         `json:"platform"`
	Enabled      bool                 `json:"enabled"`
	AccountCount int                  `json:"account_count"`
	Accounts     []SiteChannelAccount `json:"accounts"`
}

type SiteChannelAccount struct {
	SiteID         int                `json:"site_id"`
	AccountID      int                `json:"account_id"`
	AccountName    string             `json:"account_name"`
	Enabled        bool               `json:"enabled"`
	AutoSync       bool               `json:"auto_sync"`
	GroupCount     int                `json:"group_count"`
	ModelCount     int                `json:"model_count"`
	Groups         []SiteChannelGroup `json:"groups"`
	RouteSummaries []SiteRouteSummary `json:"route_summaries"`
}

type SiteRouteSummary struct {
	RouteType SiteModelRouteType `json:"route_type"`
	Count     int                `json:"count"`
}

type SiteChannelGroup struct {
	GroupKey              string             `json:"group_key"`
	GroupName             string             `json:"group_name"`
	KeyCount              int                `json:"key_count"`
	EnabledKeyCount       int                `json:"enabled_key_count"`
	MaskedPendingKeyCount int                `json:"masked_pending_key_count"`
	HasKeys               bool               `json:"has_keys"`
	HasProjectedChannel   bool               `json:"has_projected_channel"`
	ProjectedChannelIDs   []int              `json:"projected_channel_ids"`
	SourceKeys            []SiteSourceKey    `json:"source_keys,omitempty"`
	ProjectedKeys         []SiteProjectedKey `json:"projected_keys,omitempty"`
	Models                []SiteChannelModel `json:"models"`
}

type SiteSourceKey struct {
	ID          int                  `json:"id"`
	Enabled     bool                 `json:"enabled"`
	Token       string               `json:"token"`
	TokenMasked string               `json:"token_masked"`
	Name        string               `json:"name"`
	GroupKey    string               `json:"group_key"`
	GroupName   string               `json:"group_name"`
	ValueStatus SiteTokenValueStatus `json:"value_status"`
	LastSyncAt  *int64               `json:"last_sync_at,omitempty"`
}

type SiteProjectedKey struct {
	ID               int     `json:"id"`
	ChannelID        int     `json:"channel_id"`
	ChannelName      string  `json:"channel_name"`
	Enabled          bool    `json:"enabled"`
	ChannelKey       string  `json:"channel_key"`
	ChannelKeyMasked string  `json:"channel_key_masked"`
	Remark           string  `json:"remark"`
	StatusCode       int     `json:"status_code"`
	LastUseTimeStamp int64   `json:"last_use_time_stamp"`
	TotalCost        float64 `json:"total_cost"`
}

type SiteChannelModel struct {
	ModelName          string                   `json:"model_name"`
	RouteType          SiteModelRouteType       `json:"route_type"`
	RouteSource        SiteModelRouteSource     `json:"route_source"`
	ManualOverride     bool                     `json:"manual_override"`
	Disabled           bool                     `json:"disabled"`
	ProjectedChannelID *int                     `json:"projected_channel_id,omitempty"`
	RouteMetadata      *SiteModelRouteMetadata  `json:"route_metadata,omitempty"`
	History            *SiteModelHistorySummary `json:"history,omitempty"`
}

type SiteModelHistorySummary struct {
	SuccessCount  int                      `json:"success_count"`
	FailureCount  int                      `json:"failure_count"`
	LastRequestAt *int64                   `json:"last_request_at,omitempty"`
	BucketSpan    int                      `json:"bucket_span,omitempty"` // 桶宽（秒），0 表示无数据
	Buckets       []SiteModelHistoryBucket `json:"buckets,omitempty"`
}

type SiteModelHistoryBucket struct {
	Time    int64 `json:"time"`
	Success int   `json:"success"`
	Failure int   `json:"failure"`
}

type SiteModelRouteUpdateRequest struct {
	GroupKey        string             `json:"group_key" binding:"required"`
	ModelName       string             `json:"model_name" binding:"required"`
	RouteType       SiteModelRouteType `json:"route_type" binding:"required"`
	RouteRawPayload string             `json:"route_raw_payload,omitempty"`
}

type SiteModelDisableUpdateRequest struct {
	GroupKey  string `json:"group_key" binding:"required"`
	ModelName string `json:"model_name" binding:"required"`
	Disabled  bool   `json:"disabled"`
}

type SiteChannelKeyCreateRequest struct {
	GroupKey string `json:"group_key" binding:"required"`
	Name     string `json:"name,omitempty"`
}

type SiteProjectedKeyAddRequest struct {
	Enabled    bool   `json:"enabled"`
	ChannelKey string `json:"channel_key" binding:"required"`
	Remark     string `json:"remark,omitempty"`
}

type SiteProjectedKeyUpdateItem struct {
	ID         int     `json:"id" binding:"required"`
	Enabled    *bool   `json:"enabled,omitempty"`
	ChannelKey *string `json:"channel_key,omitempty"`
	Remark     *string `json:"remark,omitempty"`
}

type SiteProjectedKeyUpdateRequest struct {
	GroupKey     string                       `json:"group_key" binding:"required"`
	KeysToAdd    []SiteProjectedKeyAddRequest `json:"keys_to_add,omitempty"`
	KeysToUpdate []SiteProjectedKeyUpdateItem `json:"keys_to_update,omitempty"`
	KeysToDelete []int                        `json:"keys_to_delete,omitempty"`
}

type SiteSourceKeyAddRequest struct {
	Enabled bool   `json:"enabled"`
	Token   string `json:"token" binding:"required"`
	Name    string `json:"name,omitempty"`
}

type SiteSourceKeyUpdateItem struct {
	ID      int     `json:"id" binding:"required"`
	Enabled *bool   `json:"enabled,omitempty"`
	Token   *string `json:"token,omitempty"`
	Name    *string `json:"name,omitempty"`
}

type SiteSourceKeyUpdateRequest struct {
	GroupKey     string                    `json:"group_key" binding:"required"`
	KeysToAdd    []SiteSourceKeyAddRequest `json:"keys_to_add,omitempty"`
	KeysToUpdate []SiteSourceKeyUpdateItem `json:"keys_to_update,omitempty"`
	KeysToDelete []int                     `json:"keys_to_delete,omitempty"`
}

File: web/src/api/endpoints/site-channel.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { apiClient } from '../client';
import { logger } from '@/lib/logger';
import type { SitePlatform } from './site';

export type SiteModelRouteType =
    | 'unknown'
    | 'openai_chat'
    | 'openai_response'
    | 'anthropic'
    | 'gemini'
    | 'volcengine'
    | 'openai_embedding';

export type SiteModelRouteSource =
    | 'sync_inferred'
    | 'manual_override'
    | 'runtime_learned'
    | 'default_assigned';

export type SiteRouteSummary = {
    route_type: SiteModelRouteType;
    count: number;
};

export type SiteModelHistoryBucket = {
    time: number;
    success: number;
    failure: number;
};

export type SiteModelHistorySummary = {
    success_count: number;
    failure_count: number;
    last_request_at?: number | null;
    bucket_span?: number;
    buckets?: SiteModelHistoryBucket[];
};

export type SiteModelRouteMetadata = {
    kind?: string;
    version?: number;
    source?: string;
    route_supported: boolean;
    route_type?: SiteModelRouteType;
    enable_groups?: string[];
    supported_endpoint_types?: string[];
    heuristic_endpoint_types?: string[];
    normalized_endpoint_types?: string[];
    unsupported_reason?: string;
};

export type SiteChannelModel = {
    model_name: string;
    route_type: SiteModelRouteType;
    route_source: SiteModelRouteSource;
    manual_override: boolean;
    disabled: boolean;
    projected_channel_id?: number | null;
    route_metadata?: SiteModelRouteMetadata | null;
    history?: SiteModelHistorySummary | null;
};

export type SiteChannelGroup = {
    group_key: string;
    group_name: string;
    key_count: number;
    enabled_key_count: number;
    masked_pending_key_count: number;
    has_keys: boolean;
    has_projected_channel: boolean;
    projected_channel_ids: number[];
    source_keys: SiteSourceKey[];
    projected_keys: SiteProjectedKey[];
    models: SiteChannelModel[];
};

export type SiteSourceKey = {
    id: number;
    enabled: boolean;
    token: string;
    token_masked: string;
    name: string;
    group_key: string;
    group_name: string;
    value_status: 'ready' | 'masked_pending';
    last_sync_at?: number | null;
};

export type SiteProjectedKey = {
    id: number;
    channel_id: number;
    channel_name: string;
    enabled: boolean;
    channel_key: string;
    channel_key_masked: string;
    remark: string;
    status_code: number;
    last_use_time_stamp: number;
    total_cost: number;
};

export type SiteChannelAccount = {
    site_id: number;
    account_id: number;
    account_name: string;
    enabled: boolean;
    auto_sync: boolean;
    group_count: number;
    model_count: number;
    groups: SiteChannelGroup[];
    route_summaries: SiteRouteSummary[];
};

export type SiteChannelCard = {
    site_id: number;
    site_name: string;
    base_url: string;
    platform: SitePlatform;
    enabled: boolean;
    account_count: number;
    accounts: SiteChannelAccount[];
};

type SiteChannelModelServer = Omit<SiteChannelModel, 'route_type' | 'route_metadata' | 'history'> & {
    route_type: string | null;
    route_metadata?: SiteModelRouteMetadata | null;
    history?: {
        success_count?: number | null;
        failure_count?: number | null;
        last_request_at?: number | null;
        bucket_span?: number | null;
        buckets?: Array<{
            time?: number | null;
            success?: number | null;
            failure?: number | null;
        }> | null;
    } | null;
};

type SiteChannelGroupServer = Omit<SiteChannelGroup, 'models' | 'projected_channel_ids' | 'source_keys' | 'projected_keys'> & {
    projected_channel_ids?: number[] | null;
    source_keys?: SiteSourceKey[] | null;
    projected_keys?: SiteProjectedKey[] | null;
    models?: SiteChannelModelServer[] | null;
};

type SiteChannelAccountServer = Omit<SiteChannelAccount, 'groups' | 'route_summaries'> & {
    groups?: SiteChannelGroupServer[] | null;
    route_summaries?: Array<{
        route_type?: string | null;
        count?: number | null;
    }> | null;
};

type SiteChannelCardServer = Omit<SiteChannelCard, 'accounts'> & {
    accounts?: SiteChannelAccountServer[] | null;
};

const SITE_MODEL_ROUTE_TYPES = new Set<SiteModelRouteType>([
    'unknown',
    'openai_chat',
    'openai_response',
    'anthropic',
    'gemini',
    'volcengine',
    'openai_embedding',
]);

function normalizeSiteModelRouteType(value: string | null | undefined): SiteModelRouteType {
    if (value && SITE_MODEL_ROUTE_TYPES.has(value as SiteModelRouteType)) {
        return value as SiteModelRouteType;
    }
    return 'unknown';
}

function normalizeSiteModelRouteMetadata(
    metadata: SiteModelRouteMetadata | null | undefined,
): SiteModelRouteMetadata | null {
    if (!metadata) return null;

    return {
        ...metadata,
        route_type: metadata.route_supported
            ? normalizeSiteModelRouteType(metadata.route_type)
            : 'unknown',
        enable_groups: metadata.enable_groups ?? [],
        supported_endpoint_types: metadata.supported_endpoint_types ?? [],
        heuristic_endpoint_types: metadata.heuristic_endpoint_types ?? [],
        normalized_endpoint_types: metadata.normalized_endpoint_types ?? [],
        unsupported_reason: metadata.unsupported_reason ?? undefined,
    };
}

function normalizeSiteModel(model: SiteChannelModelServer): SiteChannelModel {
    return {
        ...model,
        route_type: normalizeSiteModelRouteType(model.route_type),
        projected_channel_id: model.projected_channel_id ?? null,
        route_metadata: normalizeSiteModelRouteMetadata(model.route_metadata),
        history: model.history
            ? {
                success_count:
                    typeof model.history.success_count === 'number'
                        ? model.history.success_count
                        : 0,
                failure_count:
                    typeof model.history.failure_count === 'number'
                        ? model.history.failure_count
                        : 0,
                last_request_at:
                    typeof model.history.last_request_at === 'number'
                        ? model.history.last_request_at
                        : null,
                bucket_span:
                    typeof model.history.bucket_span === 'number'
                        ? model.history.bucket_span
                        : 0,
                buckets: (model.history.buckets ?? []).map((b) => ({
                    time: typeof b.time === 'number' ? b.time : 0,
                    success: typeof b.success === 'number' ? b.success : 0,
                    failure: typeof b.failure === 'number' ? b.failure : 0,
                })),
            }
            : null,
    };
}

function normalizeSiteChannelAccount(account: SiteChannelAccountServer): SiteChannelAccount {
    return {
        ...account,
        groups: (account.groups ?? []).map((group) => ({
            ...group,
            masked_pending_key_count: typeof group.masked_pending_key_count === 'number' ? group.masked_pending_key_count : 0,
            projected_channel_ids: group.projected_channel_ids ?? [],
            source_keys: (group.source_keys ?? []).map((key) => ({
                ...key,
                token: typeof key.token === 'string' ? key.token : '',
                token_masked: typeof key.token_masked === 'string' ? key.token_masked : '',
                name: typeof key.name === 'string' ? key.name : '',
                group_key: typeof key.group_key === 'string' ? key.group_key : group.group_key,
                group_name: typeof key.group_name === 'string' ? key.group_name : group.group_name,
                value_status: key.value_status === 'masked_pending' ? 'masked_pending' : 'ready',
                last_sync_at: typeof key.last_sync_at === 'number' ? key.last_sync_at : null,
            })),
            projected_keys: (group.projected_keys ?? []).map((key) => ({
                ...key,
                channel_key: typeof key.channel_key === 'string' ? key.channel_key : '',
                channel_key_masked: typeof key.channel_key_masked === 'string' ? key.channel_key_masked : '',
                remark: typeof key.remark === 'string' ? key.remark : '',
                status_code: typeof key.status_code === 'number' ? key.status_code : 0,
                last_use_time_stamp: typeof key.last_use_time_stamp === 'number' ? key.last_use_time_stamp : 0,
                total_cost: typeof key.total_cost === 'number' ? key.total_cost : 0,
            })),
            models: (group.models ?? []).map(normalizeSiteModel),
        })),
        route_summaries: (account.route_summaries ?? []).map((summary) => ({
            route_type: normalizeSiteModelRouteType(summary.route_type),
            count: typeof summary.count === 'number' ? summary.count : 0,
        })),
    };
}

function normalizeSiteChannelCard(card: SiteChannelCardServer): SiteChannelCard {
    return {
        ...card,
        accounts: (card.accounts ?? []).map(normalizeSiteChannelAccount),
    };
}

export type SiteModelRouteUpdateRequest = {
    group_key: string;
    model_name: string;
    route_type: SiteModelRouteType;
    route_raw_payload?: string;
};

export type SiteModelDisableUpdateRequest = {
    group_key: string;
    model_name: string;
    disabled: boolean;
};

export type SiteChannelModelDisabledMutationInput = {
    siteId: number;
    accountId: number;
    payload: SiteModelDisableUpdateRequest[];
};

export type SiteChannelKeyCreateRequest = {
    group_key: string;
    name?: string;
};

export type SiteSourceKeyAddRequest = {
    enabled: boolean;
    token: string;
    name?: string;
};

export type SiteSourceKeyUpdateItem = {
    id: number;
    enabled?: boolean;
    token?: string;
    name?: string;
};

export type SiteSourceKeyUpdateRequest = {
    group_key: string;
    keys_to_add?: SiteSourceKeyAddRequest[];
    keys_to_update?: SiteSourceKeyUpdateItem[];
    keys_to_delete?: number[];
};

function getAccountPath(siteId: number, accountId: number, suffix: string) {
    return '/api/v1/site-channel/' + siteId + '/account/' + accountId + suffix;
}

function replaceSiteChannelAccount(
    cards: SiteChannelCard[] | undefined,
    siteId: number,
    account: SiteChannelAccount,
) {
    if (!cards) return cards;

    return cards.map((card) => {
        if (card.site_id !== siteId) return card;

        return {
            ...card,
            accounts: card.accounts.map((item) =>
                item.account_id === account.account_id ? account : item,
            ),
        };
    });
}

function invalidateSiteChannelQueries(queryClient: ReturnType<typeof useQueryClient>) {
    queryClient.invalidateQueries({ queryKey: ['site-channel', 'list'] });
}

function invalidateSiteChannelAndRelated(queryClient: ReturnType<typeof useQueryClient>) {
    queryClient.invalidateQueries({ queryKey: ['site-channel', 'list'] });
    queryClient.invalidateQueries({ queryKey: ['channels', 'list'] });
    queryClient.invalidateQueries({ queryKey: ['models', 'channel'] });
}

export function useSiteChannelList() {
    return useQuery({
        queryKey: ['site-channel', 'list'],
        queryFn: async () => apiClient.get<SiteChannelCardServer[]>('/api/v1/site-channel/list'),
        select: (cards) => cards.map(normalizeSiteChannelCard),
        refetchInterval: 30000,
    });
}

export function useUpdateSiteChannelModelRoutes(siteId: number, accountId: number) {
    const queryClient = useQueryClient();

    return useMutation({
        mutationFn: async (payload: SiteModelRouteUpdateRequest[]) =>
            apiClient.put<SiteChannelAccountServer>(getAccountPath(siteId, accountId, '/model-routes'), payload),
        onSuccess: (account) => {
            const normalizedAccount = normalizeSiteChannelAccount(account);
            queryClient.setQueryData<SiteChannelCard[]>(['site-channel', 'list'], (cards) =>
                replaceSiteChannelAccount(cards, siteId, normalizedAccount),
            );
            invalidateSiteChannelQueries(queryClient);
        },
        onError: (error) => {
            logger.error('site channel model route update failed:', error);
        },
    });
}

export function useCreateSiteChannelKey(siteId: number, accountId: number) {
    const queryClient = useQueryClient();

    return useMutation({
        mutationFn: async (payload: SiteChannelKeyCreateRequest) =>
            apiClient.post<SiteChannelAccountServer>(getAccountPath(siteId, accountId, '/keys'), payload),
        onSuccess: (account) => {
            const normalizedAccount = normalizeSiteChannelAccount(account);
            queryClient.setQueryData<SiteChannelCard[]>(['site-channel', 'list'], (cards) =>
                replaceSiteChannelAccount(cards, siteId, normalizedAccount),
            );
            invalidateSiteChannelQueries(queryClient);
        },
        onError: (error) => {
            logger.error('site channel key create failed:', error);
        },
    });
}

export function useUpdateSiteChannelModelDisabled() {
    const queryClient = useQueryClient();

    return useMutation({
        mutationFn: async ({ siteId, accountId, payload }: SiteChannelModelDisabledMutationInput) =>
            apiClient.put<SiteChannelAccountServer>(getAccountPath(siteId, accountId, '/model-disabled'), payload),
        onSuccess: (account, variables) => {
            const normalizedAccount = normalizeSiteChannelAccount(account);
            queryClient.setQueryData<SiteChannelCard[]>(['site-channel', 'list'], (cards) =>
                replaceSiteChannelAccount(cards, variables.siteId, normalizedAccount),
            );
            invalidateSiteChannelQueries(queryClient);
        },
        onError: (error) => {
            logger.error('site channel model disabled update failed:', error);
        },
    });
}

export function useUpdateSiteSourceKeys(siteId: number, accountId: number) {
    const queryClient = useQueryClient();

    return useMutation({
        mutationFn: async (payload: SiteSourceKeyUpdateRequest) =>
            apiClient.put<SiteChannelAccountServer>(getAccountPath(siteId, accountId, '/source-keys'), payload),
        onSuccess: (account) => {
            const normalizedAccount = normalizeSiteChannelAccount(account);
            queryClient.setQueryData<SiteChannelCard[]>(['site-channel', 'list'], (cards) =>
                replaceSiteChannelAccount(cards, siteId, normalizedAccount),
            );
            invalidateSiteChannelAndRelated(queryClient);
        },
        onError: (error) => {
            logger.error('site source key update failed:', error);
        },
    });
}

export function useUpdateAnySiteSourceKeys() {
    const queryClient = useQueryClient();

    return useMutation({
        mutationFn: async ({
            siteId,
            accountId,
            payload,
        }: {
            siteId: number;
            accountId: number;
            payload: SiteSourceKeyUpdateRequest;
        }) =>
            apiClient.put<SiteChannelAccountServer>(getAccountPath(siteId, accountId, '/source-keys'), payload),
        onSuccess: (account, variables) => {
            const normalizedAccount = normalizeSiteChannelAccount(account);
            queryClient.setQueryData<SiteChannelCard[]>(['site-channel', 'list'], (cards) =>
                replaceSiteChannelAccount(cards, variables.siteId, normalizedAccount),
            );
            invalidateSiteChannelAndRelated(queryClient);
        },
        onError: (error) => {
            logger.error('site source key update failed:', error);
        },
    });
}

export function useResetSiteChannelModelRoutes(siteId: number, accountId: number) {
    const queryClient = useQueryClient();

    return useMutation({
        mutationFn: async () =>
            apiClient.post<SiteChannelAccountServer>(getAccountPath(siteId, accountId, '/model-routes/reset'), {}),
        onSuccess: (account) => {
            const normalizedAccount = normalizeSiteChannelAccount(account);
            queryClient.setQueryData<SiteChannelCard[]>(['site-channel', 'list'], (cards) =>
                replaceSiteChannelAccount(cards, siteId, normalizedAccount),
            );
            invalidateSiteChannelQueries(queryClient);
        },
        onError: (error) => {
            logger.error('site channel model route reset failed:', error);
        },
    });
}
